lib/index.ts

```diff
@@ -0,0 +1,471 @@
+/**
+ * @license
+ * Copyright 2016-2021 Balena Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as _ from 'lodash';
+
+import type { DockerVersion, ImagePushOptions } from 'dockerode';
+
+export interface EngineVersion extends DockerVersion {
+	Engine?: string;
+}
+
+export interface PullPushOptions {
+	authconfig?: {
+		username?: string;
+		password?: string;
+		auth?: string;
+		email?: string;
+		serveraddress?: string;
+		registrytoken?: string;
+	};
+	ignoreProgressErrorEvents?: boolean;
+}
+
+export type ProgressCallback = (eventObj: any) => void;
+
+const DEFAULT_PROGRESS_BAR_STEP_COUNT = 50;
+
+function tryExtractDigestHash(evt: {
+	aux?: { Digest?: string };
+	status?: any;
+}): string | undefined {
+	if (evt.aux && evt.aux.Digest) {
+		return evt.aux.Digest;
+	}
+	if (typeof evt.status === 'string') {
+		const matchPull = evt.status.match(/^Digest:\s([a-zA-Z0-9]+:[a-f0-9]+)$/);
+		if (matchPull != null) {
+			return matchPull[1];
+		}
+	}
+}
+
+function awaitRegistryStream(
```

Comment:
Removed async keyword to address https://github.com/balena-io-modules/docker-progress/pull/54#discussion_r626634086